pkg/ingester/ingester.go

```diff
@@ -208,12 +208,64 @@ func (i *Ingester) starting(ctx context.Context) error {
 		return errors.Wrap(err, "failed to start lifecycler")
 	}
 
+	i.startFlushLoops()
+
+	return nil
+}
+
+func (i *Ingester) startFlushLoops() {
 	i.flushQueuesDone.Add(i.cfg.ConcurrentFlushes)
 	for j := 0; j < i.cfg.ConcurrentFlushes; j++ {
 		i.flushQueues[j] = util.NewPriorityQueue(i.metrics.flushQueueLength)
 		go i.flushLoop(j)
 	}
+}
+
+// NewForFlusher constructs a new Ingester to be used by flusher target.
+// Compared to the 'New' method:
+//   * Always replays the WAL.
+//   * Does not start the lifecycler.
+//   * No ingester v2.
+func NewForFlusher(cfg Config, clientConfig client.Config, chunkStore ChunkStore, registerer prometheus.Registerer) (*Ingester, error) {
+	if cfg.ingesterClientFactory == nil {
+		cfg.ingesterClientFactory = client.MakeIngesterClient
+	}
+
+	i := &Ingester{
+		cfg:          cfg,
+		clientConfig: clientConfig,
+		metrics:      newIngesterMetrics(registerer, true),
+		chunkStore:   chunkStore,
+		flushQueues:  make([]*util.PriorityQueue, cfg.ConcurrentFlushes),
+	}
 
+	// Should be a noop WAL.
+	cfg.WALConfig.WALEnabled = false
+	var err error
+	i.wal, err = newWAL(cfg.WALConfig, i.userStates.cp)
```

Comment:
Is `i.userStates` set at this point?

Reply:
GOOD point, it is not. I tested this already and it didn't panic. I will pass a `nil` here as we expect a noopWAL here which has no operations.